### PR TITLE
Add VariableDeclarationNode for binding variable

### PR DESCRIPTION
**Pull request by zcai1**
Pattern matching has been introduced since Java 14, which defines a new way to declare a local variable. For example, `obj instanceof String s` declares a `String s = (String) obj` when the `instanceof` evaluates to true.

However, we currently don't have a `VariableDeclarationNode` for this declaration in the CFG. This PR adds the node so we can have a more consistent behaviour.

A simple example:
image

### CODE REVIEW

**Review by wmdietl — APPROVED** [review]
Looks good, thanks!
Can you add a short mention in the changelog, in the implementation details.

**Review by wmdietl — COMMENTED** [review]
This isn't related to your change, but maybe you understand what is going on in the CFG you show in your summary... (Can you paste the whole source code for the example?)
In `NullnessStore#52 - then` and in `NullnessStore#54` the value for `s` is `@Nullable`.
It is then odd that the value of `s` after 54 is `@NonNull`, but in in `NullnessStore#56` the value in the store is `@Nullable` again.
Isn't that a bit odd? Why doesn't the `s.toLowerCase` cause a nullness error?

Also, take this example:
````
public class Binding {
  void foo(Object o) {
    if (o instanceof String s) {
      System.out.println("s is: " + s);
    } else {
      System.out.println("s is: " + s);
    }
  }
}
````

javac complains about the use of `s` in the else branch - `s` is only bound if the instanceof test is true, in the then branch.
Note, however, that `NullnessStore#52 - else` and, even more dangerously, `NullnessStore#55` contain a value for `s`.
Maybe `s` shouldn't be declared where it is currently being declared at all?

**Comment by zcai1** [comment]
image

L1071 is the cause of this strange behaviour. The `result` from L1048 only contains one store, which is an alias of `input.getRegularStore()`. The refinement on L1071 affects "both stores" in the `result`.

My suggestion is to always use `ConditionalTransferResult` as `result` in `visitInstanceOf`, and L1060 and L1071 should only refine `result.getThenStore()`.

**Comment by wmdietl** [comment]
@zcai1 Thanks for looking into this!
How about I merge this improvement as one it is now.
Can you then follow up with a new PR that looks into the scope and value for the binding variables?

**Comment by zcai1** [comment]
@wmdietl Yes, thank you!

File: docs/CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,9 @@ Version 3.28.0-eisop2 (December ?, 2022)
 
 **Implementation details:**
 
+A `VariableDeclarationNode` will be added to the CFG for the binding variable
+in a `BindingPatternTree`.
```
Comment (wmdietl) [inline docs/CHANGELOG.md:9]:
```suggestion
A `VariableDeclarationNode` is now correctly added to the CFG for the binding variable
in a `BindingPatternTree`.
```

**Review by wmdietl — APPROVED** [review]
Thanks!

**Pull request merged**